src/game/scoring.rs

```diff
@@ -1,142 +1,346 @@
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub enum CombinationKind {
+    Nob,
+    Fifteen,
+    RankMatch,
+    Run,
+    SuitMatch,
+}
 
-    local_hand.sort_by_key(|c| c.rank());
+/// Some cribbage scoring combinations have specific names
+/// depending on how many cards are involved.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub enum CombinationName {
+    Nob,
+    Fifteen,
+    Pair,
+    RunOfThree,
+    RunOfFour,
+    FlushOfFour,
+    RunOfFive,
+    FlushOfFive,
+    RoyalPair,
+    DoubleRoyalPair,
+}
 
-    local_score += score_fifteens(&local_hand);
+/// `Combination` is a record of a single scoring combination of cards.
+#[derive(Clone, Debug)]
+pub struct Combination {
+    kind: CombinationKind,
+    name: CombinationName,
+    cards: Vec<Card>,
+    rank_info: Rank, // convenience field for pairs
+    suit_info: Suit, // convenience field for flushes
+    points: u32,
+}
 
-    local_score += score_runs(&local_hand);
+/// `Score` holds a collection of scoring combinations. `score.points()`
+/// returns the sume of the points of the `combinations`.
+#[derive(Clone, Debug)]
+pub struct Score {
+    combinations: Vec<Combination>,
+    total_score: u32 // convinient sum of all combinations.score
+}
 
-    local_score += score_same_kind(&local_hand);
+/// Calculates the score for a `hand` of four cribbage cards and the `starter`.
+///
+/// # Assumptions
+///
+/// * `hand` has four unique and valid cards
+/// * `starter` is a valid `Card` that is unique when combined with `hand`
+pub fn score_hand(hand: Hand, starter: Card, is_crib: bool) -> Score {
+    let mut vector = hand.clone();
+    vector.push(starter);
+    vector.sort(); // ordered by rank rirst
+
+    let s = nob_score(hand, starter);
+    all_combinations_of_min_size(vector, 2)
+        .filter_map(|cards| score(cards, is_crib))
+        .fold(s, |mut score, combis| tally(&mut score, combis))
 
-    local_score
 }
 
-fn score_same_kind(hand: &[Card]) -> i32 {
-    let mut local_score = 0;    
-    for cards in hand.iter().combinations(2).into_iter(){
-        if cards[0].ordinal() == cards[1].ordinal()  {
-            local_score += 2;
-        }
+/// A Nob is scored if a Jack in the `hand` matches the suit of the `starter`.
+///
+/// # Returns
+///
+/// An empty `Score` or a `Score` that contains a single `Nob` `Combination`
+pub fn nob_score(hand: Hand, starter: Card) -> Score {
+    let mut score = Score::new();
+    for c in hand {
+        if c.rank == Rank::Jack && c.suit == starter.suit {
+            score.add_combination(Combination::new(CombinationKind::Nob, vec![c, starter]));
+            return score;
+        };
     }
-
-    local_score
+    score
 }
-
-fn score_nibs(hand: &[Card], shared_card: &Card) -> i32 {
-    if hand
-        .iter()
-        .any(|card| card.suit() == shared_card.suit() && card.ordinal() == Ordinal::Jack)
-    {
-        return 1;
+#[allow(clippy::single_match)]
```

Comment:
Why not use if let?

Reply:
Looked up clippy::single_match on https://rust-lang.github.io/rust-clippy/master/. It's listed as:
**What it does** Checks for matches with a single arm where an `if let` will usually suffice.
**Why is this bad?** Just readability – `if let` nests less than a `match`.

My question is more along the lines of why suppress the single_match warning instead of updating to use `if let` as it suggests.

Reply:
good point.  fixed it.